clippy_lints/src/mem_replace.rs

```diff
@@ -67,8 +67,127 @@ declare_clippy_lint! {
     "`mem::replace(&mut _, mem::uninitialized())` or `mem::replace(&mut _, mem::zeroed())`"
 }
 
+declare_clippy_lint! {
+    /// **What it does:** Checks for `std::mem::replace` on a value of type
+    /// `T` with `T::default()`.
+    ///
+    /// **Why is this bad?** `std::mem` module already has the method `take` to
+    /// take the current value and replace it with the default value of that type.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    /// ```rust
+    /// let mut text = String::from("foo");
+    /// let replaced = std::mem::replace(&mut text, String::default());
+    /// ```
+    /// Is better expressed with:
+    /// ```rust
+    /// let mut text = String::from("foo");
+    /// let taken = std::mem::take(&mut text);
+    /// ```
+    pub MEM_REPLACE_WITH_DEFAULT,
+    nursery,
+    "replacing a value of type `T` with `T::default()` instead of using `std::mem::take`"
+}
+
 declare_lint_pass!(MemReplace =>
-    [MEM_REPLACE_OPTION_WITH_NONE, MEM_REPLACE_WITH_UNINIT]);
+    [MEM_REPLACE_OPTION_WITH_NONE, MEM_REPLACE_WITH_UNINIT, MEM_REPLACE_WITH_DEFAULT]);
+
+fn check_replace_option_with_none(cx: &LateContext<'_, '_>, expr: &'_ Expr, args: &HirVec<Expr>) {
+    if let ExprKind::Path(ref replacement_qpath) = args[1].kind {
+        // Check that second argument is `Option::None`
+        if match_qpath(replacement_qpath, &paths::OPTION_NONE) {
+            // Since this is a late pass (already type-checked),
+            // and we already know that the second argument is an
+            // `Option`, we do not need to check the first
+            // argument's type. All that's left is to get
+            // replacee's path.
+            let replaced_path = match args[0].kind {
+                ExprKind::AddrOf(BorrowKind::Ref, Mutability::Mutable, ref replaced) => {
+                    if let ExprKind::Path(QPath::Resolved(None, ref replaced_path)) = replaced.kind {
```

Comment:
match is better here.

Reply:
I didn't add this code. I just moved the `replace_option_with_none` lint code into it's own function because the `check_expr` function was getting too big. I can create a separate PR to address this.